lib/hanami/db/testing.rb

```diff
@@ -16,18 +17,128 @@ module Testing
       DATABASE_NAME_MATCHER = /_dev(elopment)?$/
       private_constant :DATABASE_NAME_MATCHER
 
-      # @api private
-      # @since 2.2.0
-      def self.database_url(url)
-        url = URI(url.to_s)
+      class << self
+        # @api private
+        # @since 2.2.0
+        def database_url(url)
+          url = parse_url(url)
+
+          case deconstruct_url(url)
+          in { scheme: "sqlite", opaque: nil, path: } unless path.nil?
+            url.path = database_filename(path)
+          in { path: String => path } if path =~ DATABASE_NAME_MATCHER
+            url.path = path.sub(DATABASE_NAME_MATCHER, DATABASE_NAME_SUFFIX)
+          in { path: String => path } unless path.end_with?(DATABASE_NAME_SUFFIX)
+            url.path << DATABASE_NAME_SUFFIX
+          else
+            # do nothing
+          end
+
+          stringify_url(url)
+        end
+
+        private
+
+        # @api private
+        # @since 2.2.0
+        def parse_url(url)
+          if url.is_a?(URI::Generic)
+            # URI#dup does not duplicate internal instance
+            # variables, making mutation dangerous.
+            URI(stringify_url(url))
+          else
+            URI(url.to_s)
+          end
+        end
+
+        # rubocop:disable Metrics/AbcSize, Metrics/PerceivedComplexity
+
+        # Work around a bug in Ruby 3.0.x that erroneously omits
+        # the '//' prefix from hierarchical URLs.
```

Comment:
Boy was this an unwelcome surprise. I guess URI is just broken for the whole 3.0.x series?

Reply:
Gosh! I couldn't even find (admittedly, via a cursory search only) a matching issue on bugs.ruby-lang.org.

Reply:
I think it's [this one](https://bugs.ruby-lang.org/issues/9912)